### PR TITLE
Redesigned countdown view

### LINKED ISSUE #1034: New Auction Countdown Timer
image

asset exported at 2x so you can see everything 0.0

### PR DESCRIPTION
**Pull request by ashfurrow**
Fixes #1034.

This is a WIP because I need to update some existing snapshot tests.

The new implementation is a little more involved because we have a design that now differs depending on ~~device~~ size classes. Before, we were assuming 39pt between units centred them "close enough", but this now uses Autolayout to lay things out.

Also, it's adaptive and will work with the new iPad multitasking. Thanks to @orta for not letting me take the easy way out :sweat_smile: 

screen shot 2016-01-19 at 3 42 09 pm

screen shot 2016-01-19 at 3 42 00 pm

screen shot 2016-01-19 at 3 42 55 pm

### CODE REVIEW

**Comment by ashfurrow** [comment]
Okay... looks like it _doesn't_ fail based on existing tests because there are none. I'll add some.

**Comment by ArtsyMobile** [comment]
&nbsp; | 1 Warning
    ------------- | ------------
        :warning: | PR is classed as Work in Progress
    

  

  


<p align="right" meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://github.com/KrauseFx/danger/">danger</a>
</p>

**Comment by ashfurrow** [comment]
Tests are passing locally, most of the time:

screen shot 2016-01-19 at 5 45 53 pm

Otherwise I get locked, like CI does. Issue is tracked here: https://github.com/artsy/eigen/issues/985

Also, sometimes danger fails, complaining of lack of a CI environment. Not sure if that's tracked anywhere, didn't see any issue on the repo.

File: Artsy/Views/Artwork/Auctions/ARCountdownView.m [hunk 1]
```diff
@@ -116,7 +188,37 @@ - (NSString *)countdownString
                                                fromDate:now
                                                  toDate:self.targetDate
                                                 options:0];
-    return [NSString stringWithFormat:@"%02d : %02d : %02d : %02d", (unsigned int)components.day, (unsigned int)components.hour, (unsigned int)components.minute, (unsigned int)components.second];
+    [self updateDays:components.day hours:components.hour mintues:components.minute seconds:components.second];
+}
+
+- (void)updateDays:(NSInteger)days hours:(NSInteger)hours mintues:(NSInteger)minutes seconds:(NSInteger)seconds
```
Comment (orta) [inline Artsy/Views/Artwork/Auctions/ARCountdownView.m]: `mintues`

File: Artsy_Tests/View_Tests/ARCountdownViewTests.m [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    expect(subject).to.haveValidSnapshot();
+
+});
+
+it(@"looks ok on a regular size class", ^{
+    ARCountdownView *subject = [[ARCountdownView alloc] init];
+    subject.targetDate = [NSDate distantPast];
+    // Need to trick it into thinking its on an iPad
+    id partialMock = [OCMockObject partialMockForObject:subject];
+    [[[partialMock stub] andReturn:[UITraitCollection traitCollectionWithHorizontalSizeClass:UIUserInterfaceSizeClassRegular]] traitCollection];
```
Comment (orta) [inline Artsy_Tests/View_Tests/ARCountdownViewTests.m]: once we hit 3-5 tests doing this, it should move to a forgeries subclass IMO

**Comment by orta** [comment]
we have an issue to remove `x-callback` and thus the tests - as iOS9 obviates the need for it.

File: Artsy_Tests/View_Tests/ARCountdownViewTests.m (hunk 2 above)
Comment (ashfurrow) [inline Artsy_Tests/View_Tests/ARCountdownViewTests.m]: Agreed 💯 I opened an issue https://github.com/ashfurrow/Forgeries/issues/13

**Comment by ashfurrow** [comment]
Corrected the typo; not sure what you mean about `x-callback`.

**Comment by ashfurrow** [comment]
So... where do we go on this? Ci is failing, but it's failing on master and the failure seems unrelated to the PR. Suggestions?

**Comment by orta** [comment]
Hrm, well you mentioned to me that it was the x-callback tests, so I removed them, and the feature in #1046 - as well as updating Danger. Should see more greens now.

**Comment by ashfurrow** [comment]
AH I see! OK, once https://github.com/artsy/eigen/pull/1046 is merged, I'll pull from master. Thanks Orta :bow:

**Comment by orta** [comment]
:wine_glass:

**Comment by ashfurrow** [comment]
:green_apple:

File: Artsy/Views/Artwork/Auctions/ARCountdownView.m [hunk 3]
```diff
@@ -24,61 +35,126 @@ - (instancetype)init
...
+
+    NSArray<UILabel *> *valueLabels = @[ self.daysValueLabel, self.hoursValueLabel, self.minutesValueLabel, self.secondsValueLabel ];
+
+    [valueLabels each:^(id label) {
+        [self addSubview:label];
+        [label constrainTopSpaceToView:self.headingLabel predicate:interitemSpacing];
+    }];
+
+    // Add colons and inter-item constraints.
+    [valueLabels betweenObjects:^(UIView *lhs, UIView *rhs) {
```
Comment (alloy) [inline Artsy/Views/Artwork/Auctions/ARCountdownView.m]: Nice :+1:

File: Artsy/Views/Artwork/Auctions/ARCountdownView.m [hunk 4]
```diff
@@ -24,61 +35,126 @@ - (instancetype)init
...
+    // Constrain first/last labels to self for horizontal intrinsic width.
+    [[valueLabels firstObject] alignLeadingEdgeWithView:self predicate:@"0"];
+    [[valueLabels lastObject] alignTrailingEdgeWithView:self predicate:@"0"];
+    [UIView equalWidthForViews:valueLabels];
+
+
+    // Add unit labels under the number labels
+    NSArray<NSString *> *units = [@[ @"Days", @"Hrs", @"Min", @"Sec" ] map:^id(id string) {
+        return [string uppercaseString];
+    }];
```
Comment (alloy) [inline Artsy/Views/Artwork/Auctions/ARCountdownView.m]: Not sure I understand the point of this, can’t you just have an array of uppercase strings?

Comment (ashfurrow) [inline Artsy/Views/Artwork/Auctions/ARCountdownView.m]: Yeah, it was originally mapped inside a larger block, and I moved it out to convey better intent. I'm fine moving the an array of uppercase strings, whatever you prefer.

File: Artsy/Views/Auction/AuctionBannerView.swift [hunk 5]
```diff
@@ -33,15 +34,19 @@ class AuctionBannerView: UIView {
             countdownView.stopTimer()
         }
     }
+
+    override func traitCollectionDidChange(previousTraitCollection: UITraitCollection?) {
+        super.traitCollectionDidChange(previousTraitCollection)
+
+        // Remove all subviews and call setupViews() again to start from scratch.
+        subviews.forEach(apply(UIView.removeFromSuperview))
```
Comment (alloy) [inline Artsy/Views/Auction/AuctionBannerView.swift]: Can’t you just do something like `subviews.forEach { $0.removeFromSuperview }` ?

Comment (ashfurrow) [inline Artsy/Views/Auction/AuctionBannerView.swift]: That also works. This is a stylistic choice, I'm fine either way.

File: Artsy/Views/Artwork/Auctions/ARCountdownView.m (hunk 4 above)
Comment (alloy) [inline Artsy/Views/Artwork/Auctions/ARCountdownView.m]: Yeah I’d prefer that. The intent of this is much clearer imo:

```objc
@[ @"DAYS", @"HRS", @"MIN", @"SEC" ]
```

**Comment by alloy** [comment]
Looks great 👍

**Comment by ashfurrow** [comment]
:bow:

**Pull request merged**